test/factory/smartWalletFactory.test.ts

```diff
@@ -1,497 +1,557 @@
+        const signature = createValidPersonalSignSignature(
+          privateKey,
+          dataToSign
+        );
+
+        await smartWalletFactory.createUserSmartWallet(
+          owner.address,
+          recoverer,
+          index,
+          signature
+        );
+
+        const smartWallet = await ethers.getContractAt(
+          'SmartWallet',
+          smartWalletAddress
+        );
+
+        await expect(smartWallet.isInitialized()).to.eventually.be.true;
+      });
+
+      it('should fail with a ZERO owner address parameter', async function () {
+        const dataToSign = utils.solidityKeccak256(
+          ['address', 'address', 'address', 'uint256'],
+          [smartWalletFactory.address, constants.AddressZero, recoverer, index]
+        );
+
+        const privateKey = Buffer.from(
+          owner.privateKey.substring(2, 66),
+          'hex'
+        );
+
+        const signature = createValidPersonalSignSignature(
+          privateKey,
+          dataToSign
+        );
+
+        await expect(
+          smartWalletFactory.createUserSmartWallet(
+            owner.address,
+            recoverer,
+            index,
+            signature
+          )
+        ).to.be.rejectedWith('Invalid signature');
+      });
+
+      it('should fail when signature does not match', async function () {
+        const dataToSign = utils.solidityKeccak256(
+          ['address', 'address', 'address', 'uint256'],
+          [smartWalletFactory.address, owner.address, recoverer, index]
+        );
+
+        const privateKey = Buffer.from(
+          owner.privateKey.substring(2, 66),
+          'hex'
+        );
+
+        const signature = createValidPersonalSignSignature(
+          privateKey,
+          dataToSign
+        );
+
+        const otherAccount = Wallet.createRandom();
+
+        await expect(
+          smartWalletFactory.createUserSmartWallet(
+            otherAccount.address,
+            recoverer,
+            index,
+            signature
+          )
+        ).to.be.rejectedWith('Invalid signature');
+      });
+    });
+
+    describe('relayedUserSmartWalletCreation', function () {
+      let recoverer: string;
+      let index: number;
+      let smartWalletAddress: string;
+      let worker: SignerWithAddress;
+      let otherCaller: SignerWithAddress;
+      let token: UtilToken;
+
+      beforeEach(async function () {
+        recoverer = constants.AddressZero;
+        index = nextIndex();
+        smartWalletAddress = await smartWalletFactory.getSmartWalletAddress(
+          owner.address,
+          recoverer,
+          index
+        );
+        [worker, otherCaller] = await ethers.getSigners();
+        token = await deployContract('UtilToken');
+      });
+
+      it('should initialize the smart wallet in the expected address without paying fee', async function () {
+        const initialWorkerBalance = await token.balanceOf(worker.address);
+
+        expect(initialWorkerBalance).to.be.equal(0);
```

Comment:
Instead of performing this check, I'd rather check that the final balance is equal to the initial one. No need for them to be always zero, do they?

Reply:
updated